tests/test_library.py

```diff
@@ -0,0 +1,659 @@
+    session = backend._session
+    playlist_2 = mocker.Mock()
+    session.mixes.return_value = [playlist_2]
+    assert not tlp.browse("tidal:mix:1")
+    session.mixes.assert_called_once_with()
+
+
+def test_specific_artist_old_api(tlp, mocker, tidal_tracks, tidal_albums):
+    tlp, backend = tlp
+    session = backend._session
+    session.get_artist_top_tracks.return_value = tidal_tracks
+    session.get_artist_albums.return_value = tidal_albums
+    assert tlp.browse("tidal:artist:1") == [
+        Ref(name="Album-0", type="album", uri="tidal:album:0"),
+        Ref(name="Album-1", type="album", uri="tidal:album:1"),
+        Ref(name="Track-0", type="track", uri="tidal:track:0:0:0"),
+        Ref(name="Track-1", type="track", uri="tidal:track:1:1:1"),
+    ]
+
+    session.get_artist_top_tracks.assert_called_once_with("1")
+    session.get_artist_albums.assert_called_once_with("1")
+
+
+def test_specific_artist_new_api(tlp, mocker, tidal_albums, tidal_artists):
+    tlp, backend = tlp
+    session = backend._session
+    session.mock_add_spec(("artist",))
+    artist = tidal_artists[0]
+    artist.get_albums.return_value = tidal_albums
+    session.artist.return_value = artist
+    assert tlp.browse("tidal:artist:1") == [
+        Ref(name="Album-0", type="album", uri="tidal:album:0"),
+        Ref(name="Album-1", type="album", uri="tidal:album:1"),
+        Ref(name="Track-100", type="track", uri="tidal:track:0:7:100"),
+    ]
+    artist.get_top_tracks.assert_called_once_with()
+    artist.get_albums.assert_called_once_with()
+    session.artist.assert_has_calls([mocker.call("1"), mocker.call("1")])
+
+
+def test_lookup_no_uris(tlp, mocker):
+    tlp, backend = tlp
+    with pytest.raises(Exception):  # just check it raises
+        tlp.lookup()
+    with pytest.raises(Exception):
+        tlp.lookup("")
+    with pytest.raises(Exception):
+        tlp.lookup("somethingwhichisntauri")
+    assert not tlp.lookup("tidal:nonsuch:11")
+
+
+def test_lookup_http_error(tlp, mocker):
+    tlp, backend = tlp
+    session = backend._session
+    session.get_album_tracks.side_effect = HTTPError
+    assert not tlp.lookup("tidal:track:0:1:0")
+
+
+def test_lookup_track(tlp, mocker, tidal_tracks, compare):
+    tlp, backend = tlp
+    session = backend._session
+    session.get_album_tracks.return_value = tidal_tracks
+    res = tlp.lookup("tidal:track:0:1:0")
+    compare(tidal_tracks[:1], res, "track")
+    session.get_album_tracks.assert_called_once_with("1")
+
+
+def test_lookup_track_cached(tlp, mocker, tidal_tracks, compare):
+    tlp, backend = tlp
+    session = backend._session
+    session.get_album_tracks.return_value = tidal_tracks
+    res = tlp.lookup("tidal:track:0:1:0")
+    compare(tidal_tracks[:1], res, "track")
+    res2 = tlp.lookup("tidal:track:0:1:0")
+    assert res2 == res
+    session.get_album_tracks.assert_called_once_with("1")
+
+
+@pytest.mark.xfail  # Should we cache the album when we look up a track?
```

Comment:
I don't think we normally cache the whole album when looking up a single track.

Reply:
Yeah, this test doesn't currently pass (the `xfail` means 'expected failure').  Since we have to do an album lookup when looking up a track we *could* cache it, to speed up later track lookups for the same album.  Or this might be overkill.

Reply:
Caching an album when looking up a single track sounds OK in theory but I fear that caching a playlist would result in many more albums needing to be cached. 

What do you think? @BlackLight